### PR TITLE
Cleanup temp directories, assigning initial terminology to new segments, and added slicelet

### LINKED ISSUE #116: Remove temporary files from DICOM plugins
- [x] DICOMParametricMapPlugin
- [x] DICOMSegmentationPlugin
- [x] DICOMTID1500Plugin
- [x] copy the DICOM files created by the converters to the DICOM DB directory

### PR DESCRIPTION
**Pull request by che85**
fixes #116 
addresses #104

@fedorov please test this. I hope that fixes the the issues regarding the missing files after restarting Slicer, even when I am not sure, why I cannot reproduce it properly.

### CODE REVIEW

File: Py/QuantitativeReporting.py [hunk 1]
```diff
@@ -317,6 +320,31 @@ def createNewSegmentationNode(self):
...
+      loadedTermNames.GetValue(int(loadedTermNames.GetNumberOfValues()-2)),
+      defaults[0], defaults[1], defaults[2], "", "", "",
+      defaults[0], defaults[1], defaults[2],
+      loadedAnatNames.GetValue(int(loadedAnatNames.GetNumberOfValues()-1)),
+      "", "", "", "", "", "")
+    segment.SetTag(vtkSegmentationCore.vtkSegment.GetTerminologyEntryTagName(),
+                   segmentTerminologyTag)
+
+    # TODO: if one segment 'Tissue_2' has been deleted and 'Tissue_1' and 'Tissue_3' are the only ones left and
+    # adding a new segment we have twice 'Tissue_3'
```
Comment (fedorov) [inline Py/QuantitativeReporting.py]: It's ok - they can be edited

**Comment by fedorov** [comment]
I first tested with an older nightly, and default terminology was not populated (some complaint in the popup). with today's nightly, I get the error below:

image

Do you get this error with the nightly?

**Comment by fedorov** [comment]
That error was while trying to load an SR with plugin. While creating a new report and adding segments, I am still getting errors for initializing terminology for new segment:

image

In the terminal, I am getting this:

image

I confirm I am testing https://github.com/QIICR/QuantitativeReporting/pull/122/commits/aae5c30bbb6f3a7e734319c484cf51d41f5bd7d1

**Comment by che85** [comment]
@fedorov I tested that with my nightly build from 11/26 but I am just downloading the most recent nightly for testing and fixing.

**Comment by che85** [comment]
@cpinter Is it possible to get the Color of a category/type from a specific context?

@fedorov It should work now. There were some changes on the core...

**Comment by cpinter** [comment]
qSlicerTerminologyNavigatorWidget::recommendedColorFromTerminology(vtkSlicerTerminologyEntry*)
Is this what you mean?

I suggest reading through qSlicerTerminologyNavigatorWidget.h to get an idea about the available utility functions.

**Comment by che85** [comment]
@cpinter qSlicerTerminologyNavigatorWidget::recommendedColorFromTerminology is static but not  Q_INVOKABLE

**Comment by cpinter** [comment]
Do you need Q_INVOKABLE for other methods as well? Please take a look and I'll add them.

**Comment by che85** [comment]
I think that should be all for now.

**Comment by cpinter** [comment]
Committed

**Comment by che85** [comment]
@fedorov Please test and me know if everything works.

**Comment by fedorov** [comment]
I didn't forget, just few a bit more urgent issues I need to work on.

**Comment by che85** [comment]
No rush, I am cleaning up SliceTracker a bit

**Comment by fedorov** [comment]
It looks very nice, with few suggestions:

 - when user changes terminology for segment, I think it would be more intuitive if the segment name was automatically updated to correspond to the new tissue property type
 - what do you think about popping up Terminology dialog when user requests a new segment creation, with tissue/tissue pre-selected in the dialog?

**Comment by che85** [comment]
> when user changes terminology for segment, I think it would be more intuitive if the segment name was automatically updated to correspond to the new tissue property type

> what do you think about popping up Terminology dialog when user requests a new segment creation, with tissue/tissue pre-selected in the dialog?

I can implement that.

To me the popup makes sense and would only be one more click. The only drawback would be that we force the user to accept (select a terminology), which could also become annoying. In my opinion there should be a checkbox "don't display again" or something similar like "set as default".

**Comment by fedorov** [comment]
> In my opinion there should be a checkbox "don't display again" or something similar like "set as default".

Yes, makes sense. I agree. Bit this would need to be set for the component outside of the module, so potentially problematic.

**Comment by cpinter** [comment]
The popup with the default selection checkbox makes sense to me. The default selection should be per-segmentation I think, not application-wide.
What should happen with the color if the default is on? Consecutive colors from generic anatomy?

**Comment by che85** [comment]
That sounds reasonable to me. 

@cpinter Another question that I have. Is there any event other than ModifiedEvent, that fires when a terminology has been set? I am asking that for handling segment names in our module which should be adapted to the terminology once that has been set. The segment name would only be automatically set when the terminology changes.

**Comment by cpinter** [comment]
Currently only the Modified event is invoked (see vtkSegment::SetTag).
As terminology is stored in a tag, a TagAdded or TagModified (or similar) event could be the custom event that is fired without changing the underlying data structure. Or we could use a special member to store terminology, in which case a completely specific TerminologyChanged event could be used.

**Comment by che85** [comment]
That would make sense to me.

**Comment by cpinter** [comment]
I described two options. One is less specific (woud invoke with any tag setting) but much easier to implement.
@lassoan What do you think?

**Comment by lassoan** [comment]
It would be difficult to add meaningful support for terminology in the low-level vtkSegment class. I would recommend to keep terminology info storage in tag(s) and add a TagModified custom event.

We'll have to implement default segment name and color support in the Slicer core anyway, so I think it's better not to spend time with a workaround but just let Csaba implement it in the core. The idea is to have two flags in vtkSegment that indicate if segment name and color have been automatically generated / the user specified them explicitly. If they are automatically generated then they are automatically updated when user selects a different terminology item.

@cpinter By when do you think this default segment name&color support could be implemented?

**Comment by cpinter** [comment]
Even tomorrow if needed. @fedorov Please set a deadline for me and I'll make sure I meet it!

**Comment by fedorov** [comment]
@cpinter - there is no urgency, I think it will be good to resolve these small issues before the project week, but there is no hard deadline, now that RSNA is over. But if you would like me to set a deadline -early next week would be good. I think the idea that Andras suggests is reasonable. Let's make sure those flags are accessible from python! ;)

**Comment by che85** [comment]
I added a slicelet for quantitative reporting

image

**Comment by cpinter** [comment]
We have discussed the terminology widget with Andras, and we came up with the following:
* Name can be changed from terminology widget (since it is populated based on that selection)
* Double-clicking the name would gring up the terminology selector, instead of directly editing the custom name (because both name and terminology answers the 'what is this' question, as opposed to color which is mostly a display property). Color would be changed with terminology eslection as before. One question here whether we pop up the terminology selector when double-clicking the color, or should it bring up a color selector for custom color?

We imagined the terminology selector this way:
image

What do you think?
I'll start implementing the changes after lunch if we can agree on the design and behavior.

**Comment by fedorov** [comment]
> We imagined the terminology selector this way

Two suggestions:
* I would change the terminology selector width to extend over category/type lists only, since anatomic region context is controlled independently. 
* Earlier we discussed a checkbox to control whether terminology navigator would show up every time a new segment is added, or user is fine with just populating the default terminology/color values. Do you plan to add it? I don't see it in the redesigned widget.

Other than that, the design looks good.

> Double-clicking the name would bring up the terminology selector

I think it is rather important to be able to edit segment name without having to change the terminology for the segment. I would still prefer a separate button to invoke terminology navigator, rather than having to double-click on the segment name. 

> One question is whether we pop up the terminology selector when double-clicking the color, or should it bring up a color selector for custom color?

I think it is be more intuitive if clicking the color icon brings up color selection widget.

**Comment by cpinter** [comment]
> I would change the terminology selector width to extend over category/type lists only, since anatomic region context is controlled independently
Sorry but I don't understand this fully. Can you please elaborate?
The anatomic region section would only show up if the selected category has showAnatomy on, otherwise the dialog would contain only the two (category and type).

If we consider terminology information necessary, then showing the terminology selector when clicking on name would make sense. We'd encourage the user to choose a meaningful terminology, but they could define custom name if necessary.
On the other hand if terminology information is optional, then changing the name immediately when double-clicking the name item is OK.
I think this is the main question.

**Comment by fedorov** [comment]
>> I would change the terminology selector width to extend over category/type lists only, since anatomic region context is controlled independently
>Sorry but I don't understand this fully. Can you please elaborate?

Here's what I meant (roughly!)

image

For the segment list, I meant to suggest the following:

image

**Comment by cpinter** [comment]
Thanks for the sketches! It makes sense, I already changed the widget. I'll implement it that way.

The segment table behavior change is independent from my current work on the default segment name, so I'll not change it yet and let it ripen a bit.
I personally think that showing terminology information in a separate cell would make it too wide (just consider how many separate fields we can have: type, modifier, anatomic region, modifier), so I'd keep it in a tooltip.

Should we move this discussion to its ticket https://github.com/QIICR/QuantitativeReporting/issues/79 ?

**Comment by fedorov** [comment]
> The segment table behavior change is independent from my current work on the default segment name, so I'll not change it yet and let it ripen a bit.
I personally think that showing terminology information in a separate cell would make it too wide (just consider how many separate fields we can have: type, modifier, anatomic region, modifier), so I'd keep it in a tooltip.
Should we move this discussion to its ticket #79 ?

Yes, sure - no urgency. I suggest we sit down at the project week to discuss options and develop consensus.

**Comment by lassoan** [comment]
A. If you want to show terminology info in the segment list and there is enough space then it could be good to have a terminology column. Terminology popup can be shown by double-clicking terminology item.

B. If there are not so much space (or terminology information is not important to show) then it's better to not show terminology column. Terminology popup can be then shown by double-clicking segment name. Note that no extra clicks are needed to change segment name using terminology popup (focus can be set by default on the segment name widget, Enter closes the popup).

Both A and B make sense, in different use cases. So, probably the behavior should be configurable.

**Comment by fedorov** [comment]
> probably the behavior should be configurable

👍

**Comment by cpinter** [comment]
Change about initial terminology committed in
https://github.com/Slicer/Slicer/commit/00f4d75d9695a97d5926fe073c48077449130829

**Comment by fedorov** [comment]
@cpinter thank you!

**Pull request merged**